src/shared/Core/Trace2.cs

```diff
@@ -449,34 +499,138 @@ public abstract class Trace2Message
     [JsonProperty("line", Order = 6)]
     public int Line { get; set; }
 
+    [JsonProperty("depth", Order = 7)]
+    public int Depth { get; set; }
+
     public abstract string ToJson();
 
     public abstract string ToNormalString();
 
+    public abstract string ToPerformanceString();
+
     protected string BuildNormalString(string message)
     {
         // The normal format uses local time rather than UTC time.
-        string time = Time.ToLocalTime().ToString("HH:mm:ss.ffffff");
+        string time = Time.ToLocalTime().ToString(NormalPerfTimeFormat);
+        string source = GetSource();
+
+        // Git's TRACE2 normal format is:
+        // [<time> SP <filename>:<line> SP+] <event-name> [[SP] <event-message>] LF
+        return $"{time} {source,-33} {Event.ToString().ToSnakeCase()} {message}";
+    }
+
+    protected string BuildPerformanceString(string message)
+    {
+        // The performance format uses local time rather than UTC time.
+        var time = Time.ToLocalTime().ToString(NormalPerfTimeFormat);
+        var source = GetSource();
+
+        var sb = new StringBuilder();
+        foreach (var component in _optionalComponents)
+        {
+            sb.Append(BuildSpan(component));
+        }
+
+        // Git's TRACE2 performance format is:
+        // [<time> SP <filename>:<line> SP+
+        //     BAR SP] d<depth> SP
+        //     BAR SP <thread-name> SP+
+        //     BAR SP <event-name> SP+
+        //     BAR SP [r<repo-id>] SP+
+        //     BAR SP [<t_abs>] SP+
+        //     BAR SP [<t_rel>] SP+
+        //     BAR SP [<category>] SP+
+        //     BAR SP DOTS* <perf-event-message>
+        //     LF
+        return $"{time} {source,-29}| d{Depth} | {Thread,-24} | {Event.ToString().ToSnakeCase(),-12} {sb} | {message}";
+    }
+
+    internal string BuildSpan(PerformanceFormatComponent component)
+    {
+        var sb = new StringBuilder();
+        var property = GetType().GetProperty(component.Name);
+        var dataSize = component.SpanSize - (component.BeginPadding + component.EndPadding);
+
+        if (property != null)
+        {
+            var propertyValue = property.GetValue(this, null)?.ToString();
+            var isTime = false;
+
+            if (double.TryParse(propertyValue, out var value))
+            {
+                // Ensure the correct number of decimal places
+                propertyValue = value.ToString("F6");
+                isTime = true;
+            }
+
+            AdjustPadding(component, propertyValue, isTime, dataSize);
+            sb.Append($"|{new string(' ', component.BeginPadding)}{propertyValue}");
```

Comment:
If we have a `StringBuilder` could we avoid mixing `new string(..)` and `$"interpolated {strings}"` and just use the various `Append*` methods instead? Avoid creating lots of new `string`s for each format operation.

Reply:
I think this comment was rendered inapplicable with all the other refactorings. Let me know if I'm mistaken though.